File: java/300/LeetCode0322.java
import java.util.HashMap;
import java.util.Map;

public class LeetCode0322 {
    public int coinChange(int[] coins, int amount) {
        Map<Integer,Integer> map = new HashMap<>();
        map.put(0, 0);
        int res = help(coins, amount, map);
        return res == Integer.MAX_VALUE ? -1 : res;
    }

    public int help(int[] coins,int amount,Map<Integer,Integer> map) {
        if (amount < 0) {
            return Integer.MAX_VALUE;
        } else if (amount == 0) {
            return 0;
        } else if (map.containsKey(amount)) {
            return map.get(amount);
        }
        int min = Integer.MAX_VALUE;
        for (int c : coins) {
            int t = help(coins, amount - c, map);
            if (t != Integer.MAX_VALUE) {
                min = Math.min(min, t + 1);
            }
        }
        map.put(amount, min);
        return min;
    }
}

File: cpp/L300/LeetCode0322.cc
#include <algorithm>
#include <vector>
using std::vector;

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        if (amount == 0) {
            return 0;
        }
        std::sort(coins.begin(), coins.end());
        vector<int> dp(amount + 1, -1);
        dp[0] = 0;
        for (long long coin : coins) {
            for (long long a = 1;a <= amount;++a) {
                for (long long i = 1;i * coin <= a;++i) {
                    long long rem = a - i * coin;
                    if (dp[rem] != -1) {
                        if (dp[a] == -1) {
                            dp[a] = i + dp[rem];
                        } else {
                            dp[a] = std::min(dp[a], static_cast<int>(i + dp[rem]));
                        }
                    }  
                }
            }
        }
        return dp[amount];
    }
};